hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -20,62 +20,162 @@
+from hypothesis.errors import InvalidArgument
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+if False:
+    from typing import Text  # noqa
+    from hypothesis.searchstrategy.strategies import SearchStrategy, Ex  # noqa
+
+
+URL_SAFE_CHARACTERS = frozenset(string.ascii_letters + string.digits + "$-_.+!*'(),")
+
+
+# This file is sourced from http://data.iana.org/TLD/tlds-alpha-by-domain.txt
+# The file contains additional information about the date that it was last updated.
+with open(
+    os.path.join(os.path.dirname(__file__), "vendor", "tlds-alpha-by-domain.txt")
+) as tld_file:
+    __header = next(tld_file)
+    assert __header.startswith("#")
+    TOP_LEVEL_DOMAINS = sorted((line.rstrip() for line in tld_file), key=len)
+TOP_LEVEL_DOMAINS.insert(0, "COM")
+
+
+class DomainNameStrategy(SearchStrategy):
+    @staticmethod
+    def clean_inputs(minimum, maximum, value, variable_name):
+        if value is None:
+            value = maximum
+        elif not isinstance(value, int):
+            raise InvalidArgument(
+                "Expected integer but %s is a %s"
+                % (variable_name, type(value).__name__)
+            )
+        elif not minimum <= value <= maximum:
+            raise InvalidArgument(
+                "Invalid value %r < %s=%r < %r"
+                % (minimum, variable_name, value, maximum)
+            )
+        return value
+
+    def __init__(self, max_length=None, max_element_length=None):
+        """
+        A strategy for :rfc:`1035` fully qualified domain names.
+
+        The upper limit for max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_length is 4, corresponding to a two letter domain
+        with a single letter sub-domain.
+        The upper limit for max_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_element_length is 1 in accordance with :rfc:`1035#section-2.3.4`
+        """
+        # https://tools.ietf.org/html/rfc1035#section-2.3.4
+
+        max_length = self.clean_inputs(4, 255, max_length, "max_length")
+        max_element_length = self.clean_inputs(
+            1, 63, max_element_length, "max_element_length"
+        )
+
+        super(DomainNameStrategy, self).__init__()
+        self.max_length = max_length
+        self.max_element_length = max_element_length
+
+        # These regular expressions are constructed to match the documented
+        # information in https://tools.ietf.org/html/rfc1035#section-2.3.1
+        # which defines the allowed syntax of a sub-domain string.
+        if self.max_element_length == 1:
+            self.label_regex = r"[a-zA-Z]"
+        elif self.max_element_length == 2:
+            self.label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
+        else:
+            max_center_chars = self.max_element_length - 2
+            self.label_regex = (
+                r"[a-zA-Z]([a-zA-Z0-9\-]{0,%s}[a-zA-Z0-9])?" % max_center_chars
```

Comment:
tiny nit: `%d` instead of `%s`? This is an `int`, right?

Reply:
Either works in this case, since we're not doing any numeric formatting.

What does matter, though I feel unbelievably pedantic mentioning this, is that we should wrap max_center_chars in a length-one tuple. As well as consistency with other code, it means that ast-based tooling can tell whats happening and refactor it if needed.

Reply:
Ah, but using a non-int with `%d` throws an error that `%s` doesn't. So even if it’s not numeric formatting, it is another level of checking.

Agree on the length-one tuple.

Reply:
@alexwlchan I'm not worried about non-int values, they are already guarded against, so `%d` would be fine, i just went with `%s` since I'm building a raw string and thats where my mind was at the time, using a feature for the first time (formatting characters in raw strings) tends to narrow my focus at times. 

@Zac-HD I might not love the 1 length tuple, but now I've used it, it has forced the code formatting around in a way that makes it possible to un-abbreviate the variable name which makes my inner urges for literate/readable code quite a lot happier with it.

Reply:
I certainly don't love it either, but we can automatically upgrade to f-strings once 3.5 is EOL 😄